src/daft-logical-plan/src/optimization/rules/reorder_joins/join_graph.rs

```diff
@@ -0,0 +1,734 @@
+                        final_name_map.insert(name.to_string(), name.to_string());
+                    }
+                    join_conds_to_resolve.push((name.to_string(), plan.clone(), false));
+                }
+                self.process_node(left, join_conds_to_resolve, final_name_map);
+                let mut ready_left = vec![];
+                for _ in 0..left_on.len() {
+                    ready_left.push(join_conds_to_resolve.pop().unwrap());
+                }
+                for r in right_on {
+                    let name = r.name();
+                    if final_name_map.get(name).is_none() {
+                        final_name_map.insert(name.to_string(), name.to_string());
+                    }
+                    join_conds_to_resolve.push((name.to_string(), plan.clone(), false));
+                }
+                self.process_node(right, join_conds_to_resolve, final_name_map);
+                let mut ready_right = vec![];
+                for _ in 0..right_on.len() {
+                    ready_right.push(join_conds_to_resolve.pop().unwrap());
+                }
+                for ((lname, lnode, ldone), (rname, rnode, rdone)) in
+                    ready_left.into_iter().zip(ready_right.into_iter())
+                {
+                    if ldone && rdone {
+                        self.edges.push(JoinEdge {
+                            l_relation_name: lname.clone(),
+                            lnode: lnode.clone(),
+                            l_final_name: final_name_map.get(&lname).unwrap().to_string(),
+                            r_relation_name: rname.clone(),
+                            rnode: rnode.clone(),
+                            r_final_name: final_name_map.get(&rname).unwrap().to_string(),
+                        });
+                    }
+                }
+            }
+            // TODO(desmond): There are potentially more reorderable nodes. For example, we can move repartitions around.
+            _ => {
+                // This is an unreorderable node. All unresolved columns coming out of this node should be marked as resolved.
+                for (name, _, done) in join_conds_to_resolve.iter_mut() {
+                    if schema_names.contains(name) {
+                        *done = true;
+                    }
+                }
+                // TODO(desmond): At this point we should perform a fresh join reorder optimization starting from this
+                // node as the root node. We can do this once we add the optimizer rule.
+            }
+        }
+    }
+
+    /// Test helper function to get the number of edges that the current graph contains.
+    #[allow(dead_code)]
+    pub(crate) fn num_edges(&self) -> usize {
+        self.edges.len()
+    }
+
+    /// Test helper function to check that all relations in this graph are connected.
+    #[allow(dead_code)]
+    pub(crate) fn fully_connected(&self) -> bool {
+        let relation_set: HashSet<_> = self
```

Comment:
I think this code can be simplified much more, since it's just a clique proble,

So we don't actually need the nodes here. We can use the `ptr` of the Arc as the unique id.
1. first construct `HashMap<ptr, Vec<ptr>>`
2. create a `HashSet<ptr>` as seen.
3. pick a node and traverse and put a ptr in the set when seen.